ja/tutorials-and-examples/bookmarks/intro.rst

```diff
@@ -1,13 +1,393 @@
+        FOREIGN KEY bookmark_key(bookmark_id) REFERENCES bookmarks(id)
+    );
+
+複合主キーを``bookmarks_tags``で使おうとして注意されると思います。
+Cakeは複合主キーをほとんどどこでもサポートしているので、もっと簡単に複数の
+アプリを一つのアプリで提供できます（マルチテナント）。
+
+テーブルとカラム名は適当に決めずに、CakePHPの
+:doc:`命名規則 </intro/conventions>`に従ったほうがいいです。
+非常に簡単に開発できるようになり、いちいち余計な設定しなくて済みます。
+Cakeは十分レガシーなDBに対応できるぐらい柔軟ですが、命名規則に従うことで、時間を節約できます。
+
+
+Database 設定
+======================
+
+次にDBがどこにありどうやって接続するのかCakeに教えます。
+殆どの場合これっきり設定はいじりません。
+
+この設定はとてもわかりやすいはずです。``Datasources.default``のなかにある 配列を in the **config/app.php** にコピペして、
+設定します。設定例は以下になります。::
+
+    return [
+        // More configuration above.
+        'Datasources' => [
+            'default' => [
+                'className' => 'Cake\Database\Connection',
+                'driver' => 'Cake\Database\Driver\Mysql',
+                'persistent' => false,
+                'host' => 'localhost',
+                'username' => 'cakephp',
+                'password' => 'AngelF00dC4k3~',
+                'database' => 'cake_bookmarks',
+                'encoding' => 'utf8',
+                'timezone' => 'UTC',
+                'cacheMetadata' => true,
+            ],
+        ],
+        // More configuration below.
+    ];
+
+一回 **config/app.php** 保存したら、DBに接続できるようになっているので、
+Cakeのホームページを見ると、'CakePHP is
+able to connect to the database' がチェックされているはずです。
+
+.. note::
+
+        
+    CakePHP　の初期設定ファイルは以下にあります。
+    **config/app.default.php**.
+
+スキャットフォールド（簡易）コードの生成
+========================
+
+
+CakePHP　の命名規則にDBが従っていれば
+:doc:`ベイクコンソール </bake/usage>` でアプリの簡単な骨格が作れます
+コマンドは以下です:
+
+    // On Windows you'll need to use bin\cake instead.
+    bin/cake bake all users
+    bin/cake bake all bookmarks
+    bin/cake bake all tags
+
+これは、コントローラー、モデル、ビュー、それにタイオするテストケース、と
+ユーザーに対してのフィクスチャー、ブックマークとタグのリソースを一気に生成します。
+サーバを止めてしまった場合、再起動して **http://localhost:8765/bookmarks**に
+アクセスして下さい。.
+
+そうすると、基本的なDBにアクセスできる動くアプリが見えるはずです。
+一度、ブックマークリストのページに行ったら、ユーザーやブックマークやタグの追加ができるはずです。
+
+
+.. note::
+
+    404　not foundになってしまったら, アパッチの mod_rewrite　モジュールがロードされているか確かめて下さい。
+
+パスワードハッシュを追加
+=======================
+
+ユーザーを作ると平文でパスワードが保存されてしまいます。セキュリティー上とても良くないので直しましょう
+
+これはまた、モデルレイヤーについて紹介する良い機会です。オブジェクトを操作するメソッドと、
+違うクラスの単一のオブジェクトを分けています。
+一つのレコードに従っている機能が一つの　「エンティティ」クラスにある場合、メソッドはテーブルクラスにあるエンティティーコレクションを操作します。
+
+例えば、パスワードハッシュは個別のレコードで行われ、エンティティーオブジェクトでビヘイビアを操作します。
+なぜなら、毎回パスワードを暗号化するときにmutator/setter　メソッドを使うからです。
+ Cake　は規約ベースのセッターメソッドをエンティーにセットします。
+**src/Model/Entity/User.php**にパスワード用のセッターを追加しましょう。
+
+    namespace App\Model\Entity;
+
+    use Cake\Auth\DefaultPasswordHasher;
+    use Cake\ORM\Entity;
+
+    class User extends Entity
+    {
+
+        // Code from bake.
+
+        protected function _setPassword($value)
+        {
+            $hasher = new DefaultPasswordHasher();
+            return $hasher->hash($value);
+        }
+    }
+    
+    
+既に作ったユーザーのパスワードを変えたら、前のパスワードに変えて、暗号化されたパスワードが
+表示される。CakePHP `bcrypt
+<http://codahale.com/how-to-safely-store-a-password/>`　でハッシュ化するのがデフォルト設定です。
```

Comment:
Replace Fullwidth space to "_" after \`bcrypt \<http://codahale.com/how-to-safely-store-a-password/\>\`